wallet/wallet_v4.tact

```diff
@@ -0,0 +1,121 @@
+interface Plugin {
+    internal requested_funds#f06c7567(query_id: Uint64);
+    internal notify_removed#e4737472(query_id: Uint64);
+    internal notify_installed#6e6f7465(query_id: Uint64);
+    internal notify_removed#64737472(query_id: Uint64);
+}
+
+contract Wallet {
+    var seqno: Uint32
+    var subwallet: Uint32
+    var pubkey: Pubkey
+    var plugins: HashSet[Plugin]
+
+    bounced default() {
+        // ignore bounced messages
+    }
+
+    internal default() {
+        // ignore simple transfers
+    }
+
+    internal request_funds#706c7567(
+        query_id: Uint64,
+        requested_amount: Coins,
+        requested_extra: ExtraCurrencies,
+        msg.sender: Plugin,
+    ) {
+        verify(self.balance() - msg.amount >= requested_amount, "Not enough funds");
+
+        send(InboundCoins)
+        msg.sender.requested_funds(
+            query_id,
+            msg.amount: requested_amount,
+            msg.extra: requested_extra,
+        );
+    }
+
+    internal remove_plugin#64737472(query_id: Uint64, msg.sender: Plugin) {
+        self.remove_plugin_from_list(msg.sender);
+
+        if msg.bounceable {
+            send(InboundCoins) msg.sender.notify_removed(query_id);
+        }
+    }
+
+    external pre_condition(body: Slice) -> Slice {
+        let sign = body.read(Signature);
+        if sign.verify(body, self.pubkey) {
+            self.accept();
+            self.seqno += 1;
+            self.commit();
+            return body;
+        } else {
+            thrown("Invalid signature");
+        }
+    }
+
+    external simple_transfer#0(messages: List[(Cell, Sendmode)]) -> Slice {
+        for (msg, flags) in messages {
+            send_raw_msg(msg, flags);
+        }
+    }
+
+    external destroy_and_install_plugin#1(
+        workchain: Int8,
+        balance: Coins,
+        state_init: Ref[StateInit],
+        body: Cell,
+    ) -> Slice {
+        let address = AddressStd.new(workchain, state_init.hash());
+
+        send(PayFee + IgnoreErrors)
+            address.send_raw(
+                body,
+                msg.amount: balance,
+                msg.state_init: state_init,
+            );
+
+        self.add_plugin(address);
+    }
+
+    external install_plugin#2(
+        plugin: Plugin,
+        amount: Coins,
+        query_id: Uint64,
+    ) -> Slice {
+        let address = AddressStd.new(workchain, state_init.hash());
+
+        self.add_plugin(address);
+
+        send(PayFee + IgnoreErrors)
+            plugin.notify_installed(
+                query_id,
+                msg.amount: balance,
+            );
+    }
+
+    external remove_plugin#3(
+        plugin: Plugin,
+        query_id: Uint64,
+    ) -> Slice {
+        self.remove_plugin_from_list(plugin);
+
+        send(PayFee + IgnoreErrors)
+            plugin.notify_removed(
+                query_id,
+            );
+    }
+
+    fn get_plugin(addr: Address) -> Option[Plugin] {
+        self.plugins.get(addr.to_std())
```

Comment:
```suggestion
        self.plugins.get(addr.normalize())
```

Maybe we need to define `Set[NormalizedAddress[Plugin]]` (whatever it means ¯\\_(ツ)_/¯) so that if we forget to `.normalize()` somewhere down in the code we'd get slapped on the wrist?

Reply:
Perhaps `Address` should be itself normalized? And when it is deserialized it should be deserialized as `MsgAddressInt` and then should be converted to the normalized representation?

Motivation here is that different address representations is a serializiation-specific detail that does not matter for the end user.

Reply:
Let's first figure out exact reasons for why it's not normalized and how exactly. I hope there is a simple solution, but we also don't want to insert normalization by default everywhere we use addresses.

OTOH, we may need to — do we have to normalize each time we do `verify(self.owner == sender)`, for instance?

Reply:
>we also don't want to insert normalization by default everywhere we use addresses

Why?

>Let's first figure out exact reasons for why it's not normalized and how exactly

https://github.com/ton-blockchain/ton/blob/master/crypto/block/block.tlb#L105-L108